### PR TITLE
Add call hierarchy support for TypeScript 3.8

### PR DESCRIPTION
**Pull request by rbuckton**
This adds Call Hierarchy support for TypeScript 3.8 and is the other half of https://github.com/microsoft/TypeScript/pull/35176.

Related: https://github.com/microsoft/vscode/issues/16110, https://github.com/microsoft/TypeScript/issues/31863

cc: @mjbvz, @jrieken

### CODE REVIEW

File: extensions/package.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
   "version": "0.0.1",
   "description": "Dependencies shared by all extensions",
   "dependencies": {
-    "typescript": "3.7.3"
+    "typescript": "3.8.0-dev.20200104"
```
Comment (rbuckton) [inline extensions/package.json]: I'm unsure what the normal procedure is for referencing newer builds of TypeScript, but the insiders build referenced here is needed for the protocol API changes in the TS Server.

File: extensions/typescript-language-features/src/utils/typeConverters.ts [hunk 2]
```diff
@@ -90,3 +97,71 @@ export namespace WorkspaceEdit {
...
+			case PConst.Kind.constructSignature: return vscode.SymbolKind.Constructor;
+			case PConst.Kind.constructorImplementation: return vscode.SymbolKind.Constructor;
+			case PConst.Kind.typeParameter: return vscode.SymbolKind.TypeParameter;
+			case PConst.Kind.string: return vscode.SymbolKind.String;
+			default: return vscode.SymbolKind.Variable;
+		}
+	}
+}
+
+export namespace CallHierarchyItem {
```
Comment (mjbvz) [inline extensions/typescript-language-features/src/utils/typeConverters.ts]: Let's move the call hierarchy specific conversations into the `callHierarchy.ts`. This file has basic conversions that are used across multiple different providers

File: extensions/package.json (hunk 1 above)
Comment (mjbvz) [reply extensions/package.json]: We only ship stable TS versions in stable VS Code. That means that we can't update this until TS 3.8 is released in mid February.

To handle cases like this, we instead add forward declarations of the required TSServer types. That way users can start using a newer TS version to try out this feature if they want. But we keep the actually shipped TS version on the current stable build

Does that make sense?

**Review by mjbvz — CHANGES_REQUESTED** [review]
Thanks for handling this!

The main change is that we can't bundle TS 3.8 with VS Code yet. Instead, let's just copy over the new TS types related to call hierarchy and keep using the TS 3.7 for everything else

**Comment by jrieken** [comment]
🕺 this is soo exciting

**Comment by rbuckton** [comment]
> We only ship stable TS versions in stable VS Code. That means that we can't update this until TS 3.8 is released in mid February.

I'll update the PR shortly, that's actually how I had it set up before creating the PR.

**Review by mjbvz — APPROVED** [review]

**Comment by mjbvz** [comment]
Thanks! We'll ship VS Code 1.42 with support for call hierarchy support but users will still need to update to TS 3.8

Once we pick up TS 3.8 with VS code 1.43, it will be available to everyone.  I'll also remove the copied over protocol types at that point

**Pull request merged**